service/storage_proxy.cc

```diff
@@ -143,6 +143,12 @@ static future<rpc::tuple<Elements..., replica::exception_variant>> encode_replic
     return make_exception_future<final_tuple_type>(std::move(eptr));
 }
 
+static inline
+const dht::token& end_token(const dht::partition_range& r) {
+    static const dht::token max_token = dht::maximum_token();
+    return r.end() ? r.end()->value().token() : max_token;
```

Comment:
The partition range bound could be exclusive.

Reply:
1. this `range -> token` logic was borrowed from the call site, [this is how](https://github.com/scylladb/scylladb/blob/master/service/storage_proxy.cc#L5414) `query_partition_key_range_concurrent` decides which nodes are responsible for the request. Not sure actually how it works, if strict predecessor of the exclusive right boundary is equal to `vnode`. For example, let the initial range passed to `query_partition_key_range` be `[1, 2)` where 2 is the successor of 1 in terms of `ring_position` order and 1 is equal to `vnode`. Then `query_ranges_to_vnodes_generator() -> [[1, 1], (1, 2)]`. I don't see the check for empty range in either `query_ranges_to_vnodes_generator` or `query_partition_key_range_concurrent`, so we will probably make a request for an empty range. Maybe we should add this check?
2. there is another bug here - if the request was mode from `query_singular`, the range will be singular, it will contain no end boundary, but it should not be treated like the end boundary is equal to the `maximum_token`, we should just take the token from the start boundary instead. I'll fix this.

Reply:
It looks like a serious bug (the site you borrowed from). Maybe I'm overlooking something (like something up the stack pre-conditioning the ranges), but I don't remember anything like that and the code shouldn't be so fragile. Please file an issue, and make the new code immune.

Reply:
I created [an issue](https://github.com/scylladb/scylladb/issues/12566).

> make the new code immune.

Probably I should check for empty range in `try_apply_fence` and do nothing in this case.

Suppose the range belongs to `vnode1`, right exclusive boundary belongs to `vnode2 > vnode1`. It's possible only if `vnode1` is a strict predecessor of the exclusive right boundary. As was shown above `query_ranges_to_vnodes_generator` should split this range in two, and the case of empty range we already handled. This means we can safely use exclusive right boundary as a representative token for the range. This resolves your original question.

Reply:
> Probably I should check for empty range in try_apply_fence and do nothing in this case.

Read code paths already contain such checks in the form of `::compat::unwrap` calls, this guarantees that the range in `try_apply_fence` is really non-wrapping.

Reply:
> I created https://github.com/scylladb/scylladb/issues/12566.

I think @avikivity was not concerned with empty ranges, but with ranges which have an exclusive end bound (but are not empty).

Reply:
> I think @avikivity was not concerned with empty ranges, but with ranges which have an exclusive end bound (which are not necessarily empty).

`query_ranges_to_vnodes_generator` turns such ranges into a sequence of ranges, where the last element could be an empty range. We'll make a request for and empty range in this case, I don't see any other negative consequences of this.